netket/optimizer/qgt/qgt_jacobian_pytree_logic.py

```diff
@@ -0,0 +1,224 @@
+
+    Args:
+        forward_fn: the log wavefunction, partialed with any model state, limited to single sample, taking a split real-imaginary PyTree
+        params : a pytree of parameters p, real-imaginary split
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree"""
+    y, vjp_fun = jax.vjp(forward_fn, params, samples)
+    gr, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+    gi, _ = vjp_fun(np.array(-1.0j, dtype=jnp.result_type(y)))
+    return gr, gi
+
+
+def vmap_grad_centered_cplx(
+    forward_fn: Callable, params: PyTree, samples: Array
+) -> PyTree:
+    """Calculates centred Jacobian (i.e., subtracts MPI mean from vmap_grad)"""
+    gr, gi = vmap_grad_rc(forward_fn, params, samples)
+    # Return the real and imaginary parts of ΔOⱼₖ stacked along the sample axis
+    # Re[S] = Re[(ΔOᵣ + i ΔOᵢ)ᴴ(ΔOᵣ + i ΔOᵢ)] = ΔOᵣᵀ ΔOᵣ + ΔOᵢᵀ ΔOᵢ = [ΔOᵣ ΔOᵢ]ᵀ [ΔOᵣ ΔOᵢ]
+    return jax.tree_multimap(
+        lambda re, im: jnp.concatenate([re, im], axis=0), sub_mean(gr), sub_mean(gi)
+    )
+
+
+def vmap_grad_centered(
+    forward_fn: Callable,
+    params: PyTree,
+    samples: Array,
+    model_state: Optional[PyTree],
+    mode: str,
+) -> PyTree:
+    """
+    compute the jacobian of forward_fn(params, samples) w.r.t params
+    as a pytree using vmapped gradients for efficiency
+
+    mode can be 'real', 'complex', 'holomorphic'
+    TODO: add support for splitting complex pytree leaves into real and imag parts so there be a good default mode
+    """
+    if mode == "real":
+        # Apply real-imaginary split
+        params, reassemble = tree_to_real(params)
+        # Preapply model state and reassemble to forward_fn, restrict to one sample
+        def f(W, σ):
+            return forward_fn(
+                {"params": reassemble(W), **model_state}, σ[jnp.newaxis, :]
+            )[0]
+
+        return vmap_grad_centered_real_holo(f, params, samples)
+    elif mode == "complex":
+        # Apply real-imaginary split
+        params, reassemble = tree_to_real(params)
+        # Preapply model state and reassemble to forward_fn, restrict to one sample
+        def f(W, σ):
+            return forward_fn(
+                {"params": reassemble(W), **model_state}, σ[jnp.newaxis, :]
+            )[0]
+
+        return vmap_grad_centered_cplx(f, params, samples)
+    elif mode == "holomorphic":
+        # Preapply model state and reassemble to forward_fn, restrict to one sample
+        def f(W, σ):
+            return forward_fn({"params": W, **model_state}, σ[jnp.newaxis, :])[0]
+
+        return vmap_grad_centered_real_holo(f, params, samples)
+    else:
+        raise NotImplementedError(
+            'Differentiation mode should be one of "real", "complex", "auto", or "holomorphic", got {}'.format(
+                mode
+            )
+        )
+
+
+@partial(jax.jit, static_argnums=(0, 4, 5))
+def prepare_doks(
```

Comment:
why `dOks`? those are the `Ok`, (the log derivatives) right? In general our way has been to call them `Ok`,

Reply:
it came from @inailuig, I think it stands for ΔOk because the mean is already subtracted

Reply:
its calculating ΔOₖ and I wanted to distinguish it from Oₖ
maybe `prepare_delta_oks` ?

Reply:
I had to do it all in one function, because otherwise its a mess subtracting the mean when the real and imaginary parts are already stacked